lib/logstash/inputs/gelf.rb

```diff
@@ -68,20 +74,36 @@ def initialize(params)
   public
   def register
     require 'gelfd'
+    @tcp = nil
+    @udp = nil
   end # def register
 
   public
   def run(output_queue)
     begin
-      # udp server
-      udp_listener(output_queue)
+      if @use_tcp
+        tcp_thr = Thread.new(output_queue) do |output_queue|
+          tcp_listener(output_queue)
+        end
+      end
+      if @use_udp
+        udp_thr = Thread.new(output_queue) do |output_queue|
+          udp_listener(output_queue)
+        end
+      end
```

Comment:
before we only had one thread here, so if we got an exception we'd be pretty sure that thread was killed.
Now with two, it can happen that one thread throwns an exception, we retry the block on line 98, and we're going to recreate both threads. This could lead to leaking threads on exceptions.

I think the solution here will be for each protocol to have its own `begin/rescue` block. Another alternative is to have this plugin only use tcp OR udp instead of AND.

Reply:
Done: (---> rewrote the stop method to handle both threads.
----> Reintroduced udp as default)